src/middlewared/middlewared/plugins/crypto.py

```diff
@@ -19,80 +23,212 @@
-                crypto.FILETYPE_PEM,
-                cert_certificate
+            context = SSL.Context(SSL.TLSv1_2_METHOD)
+            context.use_certificate(public_key_obj)
+            context.use_privatekey(private_key_obj)
+            context.check_privatekey()
+        except SSL.Error as e:
+            verrors.add(
+                f'{schema_name}.privatekey',
+                f'Private key does not match certificate: {e}'
             )
-        except Exception:
-            return None
-        else:
-            return certificate.digest('sha1').decode()
+
+    key_length = data.get('key_length')
+    if key_length:
+        if key_length not in [1024, 2048, 4096]:
+            verrors.add(
+                f'{schema_name}.key_length',
+                'Key length must be a valid value ( 1024, 2048, 4096 )'
+            )
+
+    signedby = data.get('signedby')
+    if signedby:
+        valid_signing_cas = await middleware.call(
+            'certificateauthority.query',
+            [
+                ('certificate', '!=', None),
+                ('privatekey', '!=', None),
+                ('certificate', '!=', ''),
+                ('privatekey', '!=', ''),
+            ],
+        )
+
+        if signedby not in [d['id'] for d in valid_signing_cas]:
+            verrors.add(
+                f'{schema_name}.signedby',
+                'Please provide a valid signing authority'
+            )
+
+
+class CertificateService(CRUDService):
+
+    class Config:
+        datastore = 'system.certificate'
+        datastore_extend = 'certificate.cert_extend'
+        datastore_prefix = 'cert_'
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.map_functions = {
+            'CERTIFICATE_CREATE_INTERNAL': self.__create_internal,
+            'CERTIFICATE_CREATE_IMPORTED': self.__create_imported_certificate,
+            'CERTIFICATE_CREATE': self.__create_certificate,
+            'CERTIFICATE_CREATE_CSR': self.__create_csr
+        }
+
+    @filterable
+    async def query(self, filters=None, options=None):
+        if options is None:
+            options = {}
+        options['extend'] = self._config.datastore_extend
+        options['prefix'] = self._config.datastore_prefix
+        return await self.middleware.call('datastore.query', self._config.datastore, filters, options)
 
     @private
     async def cert_extend(self, cert):
         """Extend certificate with some useful attributes."""
-        if cert['cert_type'] in (
-            CA_TYPE_EXISTING, CA_TYPE_INTERNAL, CA_TYPE_INTERMEDIATE
+        if cert.get('signedby'):
+            cert['signedby'] = await self.middleware.call(
+                'certificateauthority.query',
+                [('id', '=', cert['signedby']['id'])],
+                {'get': True}
+            )
+
+        # convert san to list
+        cert['san'] = cert.pop('san').split() if cert.get('san') else []
```

Comment:
cert['san'] = cert.pop('san', '').split()

seems simpler/faster : P

Reply:
And if san is None, then : P ?

Reply:
Do we allow it to be None?

then:

cert['san'] = (cert.pop('san', '') or '').split()

Reply:
yes we do, and this is definitely more neat : )